File: demo/mob_minion/src/create_mob.js
/**
 * Create Mob Module
 * 
 * Implements the Create Mob sub-epic functionality for grouping minion tokens into mobs.
 * This module handles token selection, mob creation, and confirmation.
 */

/**
 * TokenSelector - Handles selection of multiple minion tokens
 * 
 * Story: Select Multiple Tokens
 */
export class TokenSelector {
    /**
     * @param {Object} foundryTokenApi - Foundry VTT Token API dependency
     */
    constructor(foundryTokenApi) {
        this.foundryTokenApi = foundryTokenApi;
    }

    /**
     * Select multiple minion tokens from Foundry canvas
     * 
     * Scenario: Game Master selects multiple tokens successfully
     * 
     * @returns {Array<Object>} Array of selected token objects with id and actorId
     * @throws {Error} If zero tokens are selected
     */
    selectMultipleTokens() {
        const selectedTokens = this.foundryTokenApi.getSelectedTokens();
        
        if (selectedTokens.length === 0) {
            throw new Error("At least one token must be selected");
        }

        return selectedTokens.map(token => ({
            id: token.id,
            actorId: token.actorId,
            name: token.name
        }));
    }
}

/**
 * MobCreator - Handles creation of mob entities from selected tokens
 * 
 * Story: Group Tokens Into Mob
 */
export class MobCreator {
    /**
     * @param {Object} foundryTokenApi - Foundry VTT Token API dependency
     * @param {Object} foundryActorSystem - Foundry VTT Actor System dependency
     */
    constructor(foundryTokenApi, foundryActorSystem) {
        this.foundryTokenApi = foundryTokenApi;
        this.foundryActorSystem = foundryActorSystem;
    }

    /**
     * Create a new mob entity containing selected tokens
     * 
     * Scenario: Game Master groups tokens into mob successfully
     * 
     * @param {Array<Object>} selectedTokens - Array of token objects to group
     * @returns {Object} Mob entity with id, tokenIds, and actorIds
     */
    groupTokensIntoMob(selectedTokens) {
        const mobId = this._generateMobId();
        
        const tokenIds = selectedTokens.map(token => token.id);
        const actorIds = selectedTokens.map(token => token.actorId);

        // Link all tokens to mob entity via Foundry Token API
        this._linkTokensToMob(tokenIds, mobId);

        return {
            id: mobId,
            tokenIds: tokenIds,
            actorIds: actorIds,
            strategy: null
        };
    }

    /**
     * Link tokens to mob entity via Foundry Token API
     * 
     * @private
     * @param {Array<string>} tokenIds - Array of token IDs
     * @param {string} mobId - Mob entity ID
     */
    _linkTokensToMob(tokenIds, mobId) {
        tokenIds.forEach(tokenId => {
            const token = this.foundryTokenApi.getTokenById(tokenId);
            if (token) {
                token.setFlag('mob-minion', 'mobId', mobId);
            }
        });
    }

    /**
     * Generate unique mob ID
     * 
     * @private
     * @returns {string} Unique mob identifier
     */
    _generateMobId() {
        return `mob_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }
}

/**
 * MobCreationConfirmation - Handles confirmation dialog and persistence
 * 
 * Story: Display Mob Creation Confirmation
 */
export class MobCreationConfirmation {
    /**
     * @param {Object} foundryActorSystem - Foundry VTT Actor System dependency
     * @param {Object} dialogManager - Dialog management dependency
     */
    constructor(foundryActorSystem, dialogManager) {
        this.foundryActorSystem = foundryActorSystem;
        this.dialogManager = dialogManager;
    }

    /**
     * Display confirmation dialog showing mob name and token count
     * 
     * Scenario: Game Master confirms mob creation
     * 
     * @param {Object} mobEntity - Mob entity to confirm
     * @returns {Promise<boolean>} True if confirmed, false if cancelled
     */
    async displayConfirmation(mobEntity) {
        const mobName = mobEntity.name || `Mob ${mobEntity.id}`;
        const tokenCount = mobEntity.tokenIds.length;

        return await this.dialogManager.showDialog({
            title: "Confirm Mob Creation",
            content: `Create mob "${mobName}" with ${tokenCount} tokens?`,
            buttons: {
                confirm: {
                    label: "Confirm",
                    callback: () => true
                },
                cancel: {
                    label: "Cancel",
                    callback: () => false
                }
            }
        });
    }

    /**
     * Persist mob in Foundry actor system
     * 
     * @param {Object} mobEntity - Mob entity to persist
     */
    persistMob(mobEntity) {
        this.foundryActorSystem.createEmbeddedDocuments('Actor', [{
            name: mobEntity.name || `Mob ${mobEntity.id}`,
            type: 'mob',
            flags: {
                'mob-minion': {
                    mobId: mobEntity.id,
                    tokenIds: mobEntity.tokenIds,
                    actorIds: mobEntity.actorIds,
                    strategy: mobEntity.strategy
                }
            }
        }]);
    }
}

File: demo/mob_minion/src/create_mob.py
"""
Create Mob Module

Implements the Create Mob sub-epic functionality for grouping minion tokens into mobs.
This module handles token selection, mob creation, and confirmation.
"""
import uuid
from typing import List, Dict, Optional


class TokenSelector:
    """Handles selection of multiple minion tokens.
    
    Story: Select Multiple Tokens
    """
    
    def __init__(self, foundry_token_api):
        """Initialize TokenSelector with Foundry Token API dependency.
        
        Args:
            foundry_token_api: Foundry VTT Token API dependency
        """
        self.foundry_token_api = foundry_token_api
    
    def select_multiple_tokens(self) -> List[Dict]:
        """Select multiple minion tokens from Foundry canvas.
        
        Scenario: Game Master selects multiple tokens successfully
        
        Returns:
            List[Dict]: Array of selected token objects with id and actorId
            
        Raises:
            ValueError: If zero tokens are selected
        """
        selected_tokens = self.foundry_token_api.get_selected_tokens()
        
        if len(selected_tokens) == 0:
            raise ValueError("At least one token must be selected")
        
        return [
            {
                'id': token['id'],
                'actorId': token['actorId'],
                'name': token.get('name', '')
            }
            for token in selected_tokens
        ]


class MobCreator:
    """Handles creation of mob entities from selected tokens.
    
    Story: Group Tokens Into Mob
    """
    
    def __init__(self, foundry_token_api, foundry_actor_system):
        """Initialize MobCreator with dependencies.
        
        Args:
            foundry_token_api: Foundry VTT Token API dependency
            foundry_actor_system: Foundry VTT Actor System dependency
        """
        self.foundry_token_api = foundry_token_api
        self.foundry_actor_system = foundry_actor_system
    
    def group_tokens_into_mob(self, selected_tokens: List[Dict]) -> 'Mob':
        """Create a new mob entity containing selected tokens.
        
        Scenario: Game Master groups tokens into mob successfully
        
        Args:
            selected_tokens: List of token objects to group
            
        Returns:
            Mob: Mob entity with id, token_ids, and actor_ids
        """
        mob_id = str(uuid.uuid4())
        
        token_ids = [token['id'] for token in selected_tokens]
        actor_ids = [token['actorId'] for token in selected_tokens]
        
        # Link all tokens to mob entity via Foundry Token API
        self._link_tokens_to_mob(token_ids, mob_id)
        
        return Mob(mob_id, token_ids, actor_ids)
    
    def _link_tokens_to_mob(self, token_ids: List[str], mob_id: str):
        """Link tokens to mob entity via Foundry Token API.
        
        Args:
            token_ids: List of token IDs
            mob_id: Mob entity ID
        """
        for token_id in token_ids:
            token = self.foundry_token_api.get_token_by_id(token_id)
            if token:
                token.set_flag('mob-minion', 'mobId', mob_id)


class Mob:
    """Mob entity representing a collection of minions."""
    
    def __init__(self, mob_id: str, token_ids: List[str], actor_ids: List[str]):
        """Initialize Mob entity.
        
        Args:
            mob_id: Unique mob identifier
            token_ids: List of token IDs belonging to mob
            actor_ids: List of actor IDs belonging to mob
        """
        self.id = mob_id
        self.token_ids = token_ids
        self.actor_ids = actor_ids
        self.strategy = None
        self.name = None


class MobCreationConfirmation:
    """Handles confirmation dialog and persistence.
    
    Story: Display Mob Creation Confirmation
    """
    
    def __init__(self, foundry_actor_system, dialog_manager):
        """Initialize MobCreationConfirmation with dependencies.
        
        Args:
            foundry_actor_system: Foundry VTT Actor System dependency
            dialog_manager: Dialog management dependency
        """
        self.foundry_actor_system = foundry_actor_system
        self.dialog_manager = dialog_manager
    
    def display_confirmation(self, mob: Mob) -> bool:
        """Display confirmation dialog showing mob name and token count.
        
        Scenario: Game Master confirms mob creation
        
        Args:
            mob: Mob entity to confirm
            
        Returns:
            bool: True if confirmed, False if cancelled
        """
        mob_name = mob.name or f"Mob {mob.id}"
        token_count = len(mob.token_ids)
        
        return self.dialog_manager.show_dialog({
            'title': 'Confirm Mob Creation',
            'content': f'Create mob "{mob_name}" with {token_count} tokens?',
            'buttons': {
                'confirm': {
                    'label': 'Confirm',
                    'callback': lambda: True
                },
                'cancel': {
                    'label': 'Cancel',
                    'callback': lambda: False
                }
            }
        })
    
    def persist_mob(self, mob: Mob):
        """Persist mob in Foundry actor system.
        
        Args:
            mob: Mob entity to persist
        """
        self.foundry_actor_system.save_mob(mob)
    
    def discard_mob(self, mob: Mob):
        """Discard mob entity (cancel creation).
        
        Args:
            mob: Mob entity to discard
        """
        self.foundry_actor_system.delete_mob(mob.id)
